## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
orderinayear = df["order_year"].value_counts().sort_index()
#orderinayear
grouped_counts_df = df[["order_year","category","subcategory"]].value_counts().reset_index(name='order_count')
grouped_counts_df






order_year     category         subcategory  order_count
0         2023  consumables   bathroom supplies         6375
1         2023  consumables      paper products         5831
2         2023  consumables    kitchen supplies         5706
3         2023    furniture              tables         4505
4         2023    furniture              stands         4173
5         2023     software          accounting         3736
6         2023     software  project management         3498
7         2023        decor              plants         3447
8         2023        decor            wall art         3076
9         2023  consumables                misc         2870
10        2023    furniture             seating         2646
11        2023  consumables                pens         2584
12        2023        decor                misc         1380
13        2023     software           logistics         1105
14        2023  electronics             telecom          972
15        2023  electronics        audio visual          730
16        2023  electronics             cabling          653
17        2023    furniture               desks          507
18        2023  electronics           computers          466
19        2022  consumables   bathroom supplies           49
20        2022  consumables      paper products           43
21        2022  consumables    kitchen supplies           40
22        2022    furniture              tables           35
23        2022    furniture              stands           28
24        2022     software  project management           24
25        2022  consumables                misc           21
26        2022     software          accounting           20
27        2022        decor              plants           19
28        2022  consumables                pens           19
29        2022    furniture             seating           18
30        2022        decor            wall art           17
31        2022     software           logistics           17
32        2022        decor                misc           15
33        2022  electronics             telecom            6
34        2022    furniture               desks            3
35        2022  electronics        audio visual            2
36        2022  electronics             cabling            2
37        2022  electronics           computers            1

In [5]:
# What three item categories had the most entries?
grouped_categories_df = df[["category"]].value_counts().reset_index(name='order_count')
grouped_categories_df
df_consumables = df[df["category"]=="consumables"]
df_consumables


first       last                                      job  \
1        Tiffany      Myers                          Music therapist   
3         Nathan      Baker                    Accounting technician   
4      Christina   Schwartz                             Chiropractor   
6          Brian      Moore                                  Barista   
7          Jesus     Wright                       Theme park manager   
...          ...        ...                                      ...   
54629      Linda  Henderson  Scientist, research (physical sciences)   
54632     Austin     Fuller                       Investment analyst   
54633     Tamara      Jones                        Fish farm manager   
54635    Derrick      Moore                                 Musician   
54636     Monica  Gutierrez                         Graphic designer   

                     phone                                email  client_id  \
1         201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
3         827-788-8123x012               bakernathan@benson.com      46554   
4             265-829-3643  christinaschwartz9252@mcconnell.com      92089   
6      +1-893-400-2733x982          moore.official@martinez.com      87213   
7       605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
...                    ...                                  ...        ...   
54629  +1-664-405-4100x443         lhenderson@boyer-krueger.com      79817   
54632         379.662.7549           fulleraustin7494@nunez.com      50317   
54633     375.495.3124x931  tamara.jones.9326@george-carter.com      22109   
54635         358.661.5483         derrick.moore.2602@pope.info      41908   
54636     294.805.9100x339            gutierrezm3195@morris.org      35176   

       order_id  order_date  order_week  order_year         item_id  \
1       8069089  2023-05-19          20        2023  XDA18116-89-4A   
3       9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4       1322274  2023-05-28          21        2023  BZX55559-12-3X   
6       8080280  2023-04-19          16        2023  UAC50632-25-4E   
7       8946782  2023-02-14           7        2023  UEU25316-01-1U   
...         ...         ...         ...         ...             ...   
54629   6473704  2023-03-16          11        2023  ZCC45621-74-4A   
54632   5064968  2023-05-26          21        2023  CUE68543-95-3U   
54633   8568920  2023-04-26          17        2023  CDX06274-12-1E   
54635   6290153  2023-02-08           6        2023  EBD29277-23-7U   
54636   8692622  2023-03-05           9        2023  EBD30093-55-9Z   

          category        subcategory  unit_price  unit_cost  unit_weight  \
1      consumables               pens       24.95      15.09         1.49   
3      consumables               pens       36.42      24.85         1.23   
4      consumables               misc      195.10     108.17        46.43   
6      consumables   kitchen supplies      245.56     231.95         1.73   
7      consumables               pens       20.99      16.97         5.12   
...            ...                ...         ...        ...          ...   
54629  consumables   kitchen supplies       72.60      52.42         0.33   
54632  consumables     paper products      557.31     409.72         2.21   
54633  consumables   kitchen supplies        8.16       5.10         3.84   
54635  consumables   kitchen supplies      206.59     175.46        11.70   
54636  consumables  bathroom supplies       65.66      57.31         4.16   

       qty  line_number  
1       21            0  
3       29            3  
4       20            1  
6       56            5  
7       22            4  
...    ...          ...  
54629  153            1  
54632  330            1  
54633   41            2  
54635   47            0  
54636  475            6  

[23538 rows x 18 columns]

In [6]:
# For the category with the most entries, which subcategory had the most entries?
grouped_subcategories_df = df_consumables[["subcategory"]].value_counts().reset_index(name='order_count')
grouped_subcategories_df

subcategory  order_count
0  bathroom supplies         6424
1     paper products         5874
2   kitchen supplies         5746
3               misc         2891
4               pens         2603

In [7]:
# Which five clients had the most entries in the data?
df_mostclientids = df["client_id"].value_counts().reset_index(name='clientid counts')
df_mostclientids.head(5)

client_id  clientid counts
0      33615              220
1      66037              211
2      46820              209
3      38378              207
4      24741              207

In [8]:
# Store the client ids of those top 5 clients in a list.
top5clientids = df_mostclientids["client_id"].head(5).tolist()

top5clientids



[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
df_mostqtyofaclient = df[df["client_id"] == 33615]
totalunits = df_mostqtyofaclient["qty"].sum()
print(f"Total units for client 33615 = {totalunits}")


Total units for client 33615 = 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal']=df["qty"] * df["unit_price"]
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

df["shippingprice"] = df["unit_weight"].apply(lambda w: w * 7 if w > 50 else w * 10 )
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  shippingprice  
0                1  115164.00           75.0  
1                0     523.95           14.9  
2                6     527.28           16.8  
3                3    1056.18           12.3  
4                1    3902.00          464.3  
...            ...        ...            ...  
54634            8    2743.29           22.5  
54635            0    9709.73          117.0  
54636            6   31188.50           41.6  
54637            0     165.76          180.4  
54638            5    3103.31     

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["totalprice"]=( (df["subtotal"] + df["shippingprice"]) * ( 1 + 0.0925))
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number   subtotal  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       shippingprice     totalprice  
0               75.0  125898.607500  
1               14.9     588.693625  
2               16.8     594.407400  
3               12.3    1167.314400  
4              464.3    4770.182750  
...              ...            ...  
54634           22.5    3021.625575  
54635          117.0   10735.702525  
54636           41.6   34118.884250  
54637          180.4     37

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df["totalcost"]=((df["unit_cost"] * df["qty"]) + df["shippingprice"]) 
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   subtotal  shippingprice  \
0             7.50   105            1  115164.00           75.0   
1             1.49    21            0     523.95           14.9   
2             1.68    39            6     527.28           16.8   
3             1.23    29            3    1056.18           12.3   
4            46.43    20            1    3902.00          464.3   
...            ...   ...          ...        ...            ...   
54634         2.25    33            8    2743.29           22.5   
54635        11.70    47            0    9709.73          117.0   
54636         4.16   475            6   31188.50           41.6   
54637        18.04   112            0     165.76          180.4   
54638         2.07  1031            5    3103.31           20.7   

          totalprice  totalcost  
0      125898.607500   80159.55  
1         588.693625     331.79  
2         594.407400     323.34  
3        1167.314400     732.95  
4        4770.182750    2627.70  
...              ...        ...  
54634    3021.625575    1725.30  
54635   10735.702525    8363.62  
54636   34118.884250   27263.85  
54637     378.179800     317.04  
54638    3412.980925 

In [14]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = df["totalprice"] - df["totalcost"]
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shippingprice     totalprice  totalcost  \
0                1  115164.00           75.0  125898.607500   80159.55   
1                0     523.95           14.9     588.693625     331.79   
2                6     527.28           16.8     594.407400     323.34   
3                3    1056.18           12.3    1167.314400     732.95   
4                1    3902.00          464.3    4770.182750    2627.70   
...            ...        ...            ...            ...        ...   
54634            8    2743.29           22.5    3021.625575    1725.30   
54635            0    9709.73          117.0   10735.702525    8363.62   
54636            6   31188.50           41.6   34118.884250   27263.85   
54637            0     165.76          180.4     378.179800     317.04   
54638            5    3103.31           20.7    3412.980925    1680.61   

             profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637     61.139800  
54638   1732.370925  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
orderids = [2742071,2173913,6128929]

df_filtered = df[df["order_id"].isin(orderids)]
df_filtered = df_filtered.groupby("order_id")['totalprice'].sum().reset_index()
df_filtered


order_id     totalprice
0   2173913  119838.357050
1   2742071  139844.512025
2   6128929  752748.745475

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
df_topclient = df.groupby("client_id")["qty"].sum().reset_index()
df_topclient.sort_values("qty",ascending=False).head(5)



client_id      qty
758      78965  4033302
850      86721  1010426
204      30143   846069
840      85931   644423
738      77354   506382

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
df_topclientsummary = df.groupby("client_id").agg(TotalUnitsPurchased = ("qty","sum"),TotalShippingPrice = ("shippingprice","sum"),TotalRevenue=("totalprice","sum"),TotalProfit=("profit","sum"))
df_topclientsummary = df_topclientsummary.sort_values("TotalUnitsPurchased",ascending=False).head(5)
df_topclientsummary



TotalUnitsPurchased  TotalShippingPrice  TotalRevenue   TotalProfit
client_id                                                                     
78965                  4033302              7383.3  6.500970e+08  2.468595e+08
86721                  1010426              6154.6  1.109862e+08  3.702421e+07
30143                   846069              2157.3  2.351848e+07  1.041976e+07
85931                   644423              2763.0  2.263611e+08  1.083580e+08
77354                   506382              2564.3  4.592791e+07  1.022964e+07

In [18]:
# Format the data and rename the columns to names suitable for presentation.
from decimal import Decimal 

# Define the money columns. 
df_topclientsummary["TotalShippingPrice"] = df_topclientsummary["TotalShippingPrice"].apply(lambda x: Decimal(x)).map('${:,.2f}'.format)
#df_topclientsummary["TotalRevenue"] = df_topclientsummary["TotalRevenue"].apply(lambda x: Decimal(x)).map('${:,.2f}'.format)
#df_topclientsummary["TotalProfit"] = df_topclientsummary["TotalProfit"].apply(lambda x: Decimal(x)).map('${:,.2f}'.format)
#df_topclientsummary

def dollartomillions(amount):
    millions = (amount) / 100000
    return f"${millions:,.2f}M"
    

# Define a function that converts a dollar amount to millions.

# Apply the currency_format_millions function to only the money columns. 
df_topclientsummary["TotalRevenueInMillions"]=df_topclientsummary["TotalRevenue"].apply(lambda x: dollartomillions(x))
df_topclientsummary["TotalProfitInMillions"]=df_topclientsummary["TotalProfit"].apply(lambda x: dollartomillions(x))
df_topclientsummary["TotalRevenue"] = df_topclientsummary["TotalRevenue"].apply(lambda x: Decimal(x)).map('${:,.2f}'.format)
df_topclientsummary["TotalProfit"] = df_topclientsummary["TotalProfit"].apply(lambda x: Decimal(x)).map('${:,.2f}'.format)



# Rename the columns to reflect the change in the money format. 
df_topclientsummary.rename(columns = {'client_id': 'Client ID','TotalUnitsPurchased': 'Total Units Purchased','TotalShippingPrice':'Total Shipping Price','TotalRevenue':'Total Revenue','TotalProfit': 'Total Profit','TotalRevenueInMillions': 'Total Revenue (millions)','TotalProfitInMillions': 'Total Profit (millions)'},inplace=True)
df_topclientsummary



Total Units Purchased Total Shipping Price    Total Revenue  \
client_id                                                                
78965                    4033302            $7,383.30  $650,096,959.42   
86721                    1010426            $6,154.60  $110,986,238.44   
30143                     846069            $2,157.30   $23,518,475.21   
85931                     644423            $2,763.00  $226,361,135.01   
77354                     506382            $2,564.30   $45,927,908.98   

              Total Profit Total Revenue (millions) Total Profit (millions)  
client_id                                                                    
78965      $246,859,505.41               $6,500.97M              $2,468.60M  
86721       $37,024,209.98               $1,109.86M                $370.24M  
30143       $10,419,759.21                 $235.18M                $104.20M  
85931      $108,358,048.60               $2,263.61M              $1,083.58M  
77354       $10,229,638.87                 $459.28M                $102.30M

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df_topclientsummary["Total Profit (millions)"] = df_topclientsummary["Total Profit (millions)"].replace({'\$':'',',':'','M':''},regex=True).astype(float)
#df_topclientsummary
new_df = df_topclientsummary.sort_values("Total Profit (millions)",ascending=False)
new_df["Total Profit (millions)"]= new_df["Total Profit (millions)"].map('${:,.2f}M'.format)
new_df



Total Units Purchased Total Shipping Price    Total Revenue  \
client_id                                                                
78965                    4033302            $7,383.30  $650,096,959.42   
85931                     644423            $2,763.00  $226,361,135.01   
86721                    1010426            $6,154.60  $110,986,238.44   
30143                     846069            $2,157.30   $23,518,475.21   
77354                     506382            $2,564.30   $45,927,908.98   

              Total Profit Total Revenue (millions) Total Profit (millions)  
client_id                                                                    
78965      $246,859,505.41               $6,500.97M              $2,468.60M  
85931      $108,358,048.60               $2,263.61M              $1,083.58M  
86721       $37,024,209.98               $1,109.86M                $370.24M  
30143       $10,419,759.21                 $235.18M                $104.20M  
77354       $10,229,638.87                 $459.28M                $102.30M

In [20]:
'''Summary 
ClientID - 78965 made the highest profit of approx $2,470M with total revenue of $6500M
ClientID - 85931 made the second profit of approx $1080M with total revenue of $2270M. The total units purchased is far less than the client id with the highest profit
'''

'Summary \nClientID - 78965 made the highest profit of approx $2,470M with total revenue of $6500M\nClientID - 85931 made the second profit of approx $1080M with total revenue of $2270M. The total units purchased is far less than the client id with the highest profit\n'